# Segmenting and Clustering Neighborhoods in Toronto
This project explores, segments, and clusters the neighborhoods in the city of Toronto. <br><br> 
For the data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. The code scrapes the Wikipedia page, wrangles and cleans the data, and reads it into a pandas dataframe so that it is in a structured format. <br><br> 
Once the data is in a structured format, we can start utilizing the Foursquare API to explore and cluster the neighborhoods in the city of Toronto.

In [3]:
import numpy as np 
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
from pandas.io.json import json_normalize
import matplotlib.cm as cm 
import matplotlib.colors as colors
import requests 
import folium # map rendering library

# pip install geopy # uncomment this line if you haven't if you haven't install geopy 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
# pip install sklearn # uncomment this line if you haven't install sklearn   
from sklearn.cluster import KMeans

## Part 1. Scrap Postal Codes from Wikipidea
Use BeautifulSoup package to scrape the Wikipedia page and obtain the data that is in the table of postal codes.


In [4]:
# pip install bs4 # uncomment this line if you haven't if you haven't install bs4 
from bs4 import BeautifulSoup
import lxml

url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(url,'html.parser')
table_classes = soup.find('table',{'class':'wikitable sortable'})

headers = "Postcode, Borough, Neighbourhood"
toronto_table = ""
for tr in table_classes.find_all('tr'):
    row1 = ""
    for tds in tr.find_all('td'):
        row1 = row1 + "," + tds.text.strip()
    toronto_table = toronto_table + row1[1:] + '\n'
#print(toronto_table)

# writing the data into the file 'toronto_data.csv'.
file = open("toronto_data.csv","wb")
file.write(bytes(toronto_table,encoding="ascii",errors="ignore"))

6211

Convert the data into dataframe.

In [5]:
df = pd.read_csv('toronto_data.csv', header = None, names = ['PostalCode', 'Borough', 'Neighbourhood'])
df.head()

PostalCode           Borough               Neighbourhood
0        M1A      Not assigned                         NaN
1        M2A      Not assigned                         NaN
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

Ignore cells with a borough that is Not assigned.

In [6]:
indexNames = df[ df['Borough'] =='Not assigned'].index
df.drop(indexNames , inplace=True)
df.head(10)

PostalCode           Borough                                 Neighbourhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
8         M9A         Etobicoke                              Islington Avenue
9         M1B       Scarborough                               Malvern / Rouge
11        M3B        North York                                     Don Mills
12        M4B         East York              Parkview Hill / Woodbine Gardens
13        M5B  Downtown Toronto                               Garden District

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [7]:
df.loc[df['Neighbourhood'] =='Not assigned' , 'Neighbourhood'] = df['Borough']
df.head()

PostalCode           Borough                                 Neighbourhood
2        M3A        North York                                     Parkwoods
3        M4A        North York                              Victoria Village
4        M5A  Downtown Toronto                    Regent Park / Harbourfront
5        M6A        North York             Lawrence Manor / Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

####  <font color='red'>The Wikipedia page has already merged neighbourhoods based on postal code !</font> 

In [8]:
df.shape

(103, 3)

<br><br>
## Part 2. Obtain the latitude and the longitude coordinates.

Import a csv file that has the geographical coordinates of each postal code.

In [9]:
df_lon_lat = pd.read_csv('Geospatial_Coordinates.csv')
df_lon_lat.columns = ['PostalCode', 'Latitude', 'Longitude']
df_lon_lat.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

Merge both the dataframes in one single dataframe.

In [10]:
df_toronto = pd.merge(df, df_lon_lat[['PostalCode','Latitude', 'Longitude']], on='PostalCode')
df_toronto.head()

PostalCode           Borough                                 Neighbourhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park / Harbourfront   
3        M6A        North York             Lawrence Manor / Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

<br><br>
## Part 3. Explore and cluster the neighborhoods in Toronto.

### a). Create a map.

Make sure that the dataset has all 10 boroughs and 103 neighborhoods.

In [11]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_toronto['Borough'].unique()),
        df_toronto.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


Use geopy library to get the latitude and longitude values of New York City.

In [12]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto")
location = geolocator.geocode(address)
latitude_toronto = location.latitude
longitude_toronto = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude_toronto, longitude_toronto))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Create a map of Toronto with neighborhoods superimposed on top.

In [13]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=10)

# add markers to map
for lat, lng, borough, Neighbourhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighbourhood']):
    label = '{}, {}'.format(Neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### b). Explore the first neighborhood in the dataframe.

Define Foursquare Credentials and Version.

In [14]:
CLIENT_ID = '4UEA0VFQ05UQ040QWYRIMPVNFOXT5DH4UJJIXEBHP1V1HOGI' 
CLIENT_SECRET = 'JNV3JFET24ZN4XKRJDYK4AUESI2QBZ4BCDFVDALXNAINJUPE' 
VERSION = '20200425' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4UEA0VFQ05UQ040QWYRIMPVNFOXT5DH4UJJIXEBHP1V1HOGI
CLIENT_SECRET:JNV3JFET24ZN4XKRJDYK4AUESI2QBZ4BCDFVDALXNAINJUPE


Get the first neighbourhood's info.

In [15]:
neighbourhood_latitude = df_toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighbourhood_longitude = df_toronto.loc[0, 'Longitude'] # neighborhood longitude value
neighbourhood_name = df_toronto.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Get the top 100 venues that are in Parkwoods within a radius of 500 meters.

In [16]:
# Defining radius and limit of venues to get
LIMIT=100
radius=500

# Create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url 


'https://api.foursquare.com/v2/venues/explore?&client_id=4UEA0VFQ05UQ040QWYRIMPVNFOXT5DH4UJJIXEBHP1V1HOGI&client_secret=JNV3JFET24ZN4XKRJDYK4AUESI2QBZ4BCDFVDALXNAINJUPE&v=20200425&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [25]:
# Send the GET request and examine the resutls.
results = requests.get(url).json()

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Clean the json file and structure it into a pandas dataframe.

In [19]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name                  categories        lat  \
0                    Brookbanks Park                        Park  43.751976   
1                      Variety Store           Food & Drink Shop  43.751974   
2  Corrosion Service Company Limited  Construction & Landscaping  43.752432   

         lng  
0 -79.332140  
1 -79.333114  
2 -79.334661

In [20]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

3 venues were returned by Foursquare.


### c). Explore all the neighbourhoods in Toronto.

Create a function to repeat the same process to all the neighbourhoods.

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


 Call the function on each neighbourhood and create a new dataframe called toronto_venues.

In [28]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park / Harbourfront
Lawrence Manor / Lawrence Heights
Queen's Park / Ontario Provincial Government
Islington Avenue
Malvern / Rouge
Don Mills
Parkview Hill / Woodbine Gardens
Garden District
Glencairn
West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale
Rouge Hill / Port Union / Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood
Guildwood / Morningside / West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor / Wilson Heights / Downsview North
Thorncliffe Park
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Scarborough Village
Fairview / Henry Farm / Oriole
Northwood Park / York University
East Toronto
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
Kennedy Park / Ionview / East Birchmount Park
Bayview Village
Downsview
T

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                               Venue  Venue Latitude  Venue Longitude  \
0                    Brookbanks Park       43.751976       -79.332140   
1                      Variety Store       43.751974       -79.333114   
2  Corrosion Service Company Limited       43.752432       -79.334661   
3             Victoria Village Arena       43.723481       -79.315635   
4                        Tim Hortons       43.725517       -79.313103   

               Venue Category  
0                        Park  
1           Food & Drink Shop  
2  Construction & Landscaping  
3                Hockey Arena  
4                 Coffee Shop

Check the size of the dataframe.

In [29]:
print(toronto_venues.shape)
toronto_venues.head()

(2128, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                               Venue  Venue Latitude  Venue Longitude  \
0                    Brookbanks Park       43.751976       -79.332140   
1                      Variety Store       43.751974       -79.333114   
2  Corrosion Service Company Limited       43.752432       -79.334661   
3             Victoria Village Arena       43.723481       -79.315635   
4                        Tim Hortons       43.725517       -79.313103   

               Venue Category  
0                        Park  
1           Food & Drink Shop  
2  Construction & Landscaping  
3                Hockey Arena  
4                 Coffee Shop

Check how many venues were returned for each neighbourhood.

In [30]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                                5   
Alderwood / Long Branch                                                 10   
Bathurst Manor / Wilson Heights / Downsview North                       20   
Bayview Village                                                          4   
Bedford Park / Lawrence Manor East                                      25   
Berczy Park                                                             57   
Birch Cliff / Cliffside West                                             4   
Brockton / Parkdale Village / Exhibition Place                          22   
Business reply mail Processing CentrE                                   18   
CN Tower / King and Spadina / Railway Lands / H...                      18   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   12   
Cedarbrae                                                                8   
Central Bay Street                                                      63   
Christie                                                                18   
Church and Wellesley                                                    72   
Clarks Corners / Tam O'Shanter / Sullivan                               15   
Cliffside / Cliffcrest / Scarborough Village West                        2   
Commerce Court / Victoria Hotel                                        100   
Davisville                                                              33   
Davisville North                                                         6   
Del Ray / Mount Dennis / Keelsdale and Silverthorn                       4   
Don Mills                                                               28   
Dorset Park / Wexford Heights / Scarborough Tow...                       5   
Downsview                                                               17   
Dufferin / Dovercourt Village                                           16   
East Toronto                                                             3   
Eringate / Bloordale Gardens / Old Burnhamthorp...                       8   
Fairview / Henry Farm / Oriole                                          60   
First Canadian Place / Underground city                                100   
Forest Hill North & West                                                 5   
Garden District                                                        100   
Glencairn                                                                4   
Golden Mile / Clairlea / Oakridge                                        9   
Guildwood / Morningside / West Hill                                      7   
Harbourfront East / Union Station / Toronto Isl...                     100   
High Park / The Junction South                                          22   
Hillcrest Village                                                        4   
Humber Summit                                                            2   
Humberlea / Emery                                                        2   
Humewood-Cedarvale                                                       3   
India Bazaar / The Beaches West                                         18   
Kennedy Park / Ionview / East Birchmount Park                            6   
Kensington Market / Chinatown / Grange Park                             58   
Kingsview Village / St. Phillips / Martin Grove...                       4   
Lawrence Manor / Lawrence Heights                                       13   
Lawrence Park                                                            4   
Leaside                                                                 33   
Little Portugal / Trinity                                               41   
Malvern / Rouge                                                          1

Check how many unique categories can be curated from all the returned venues.

In [31]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 268 uniques categories.


### d). Analyze Each Neighborhood.

In [32]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot.head()

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Business Service  \
0             0          

In [33]:
toronto_onehot.shape

(2128, 269)

Group rows by neighbourhood and by taking the mean of the frequency of occurrence of each category.

In [34]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                           Agincourt           0.000000   
1                             Alderwood / Long Branch           0.000000   
2   Bathurst Manor / Wilson Heights / Downsview North           0.000000   
3                                     Bayview Village           0.000000   
4                  Bedford Park / Lawrence Manor East           0.000000   
5                                         Berczy Park           0.000000   
6                        Birch Cliff / Cliffside West           0.000000   
7      Brockton / Parkdale Village / Exhibition Place           0.000000   
8               Business reply mail Processing CentrE           0.000000   
9   CN Tower / King and Spadina / Railway Lands / ...           0.000000   
10                                Caledonia-Fairbanks           0.000000   
11              Canada Post Gateway Processing Centre           0.000000   
12                                          Cedarbrae           0.000000   
13                                 Central Bay Street           0.000000   
14                                           Christie           0.000000   
15                               Church and Wellesley           0.000000   
16          Clarks Corners / Tam O'Shanter / Sullivan           0.000000   
17  Cliffside / Cliffcrest / Scarborough Village West           0.000000   
18                    Commerce Court / Victoria Hotel           0.000000   
19                                         Davisville           0.000000   
20                                   Davisville North           0.000000   
21  Del Ray / Mount Dennis / Keelsdale and Silvert...           0.000000   
22                                          Don Mills           0.000000   
23  Dorset Park / Wexford Heights / Scarborough To...           0.000000   
24                                          Downsview           0.000000   
25                      Dufferin / Dovercourt Village           0.000000   
26                                       East Toronto           0.000000   
27  Eringate / Bloordale Gardens / Old Burnhamthor...           0.000000   
28                     Fairview / Henry Farm / Oriole           0.000000   
29            First Canadian Place / Underground city           0.000000   
30                           Forest Hill North & West           0.000000   
31                                    Garden District           0.000000   
32                                          Glencairn           0.000000   
33                  Golden Mile / Clairlea / Oakridge           0.000000   
34                Guildwood / Morningside / West Hill           0.000000   
35  Harbourfront East / Union Station / Toronto Is...           0.000000   
36                     High Park / The Junction South           0.000000   
37                                  Hillcrest Village           0.000000   
38                                      Humber Summit           0.000000   
39                                  Humberlea / Emery           0.000000   
40                                 Humewood-Cedarvale           0.000000   
41                    India Bazaar / The Beaches West           0.000000   
42      Kennedy Park / Ionview / East Birchmount Park           0.000000   
43        Kensington Market / Chinatown / Grange Park           0.000000   
44  Kingsview Village / St. Phillips / Martin Grov...           0.000000   
45                  Lawrence Manor / Lawrence Heights           0.076923   
46                                      Lawrence Park           0.000000   
47                                            Leaside           0.000000   
48                          Little Portugal / Trinity           0.000000   
49                                    Malvern / Rouge           0.000000   
50  Milliken / Agincourt North / Steeles East / L'...           0.000000   
51  Mimico NW / The Queensway West / South of Bloo...           0.000000   
52          

Confirm the new size.

In [35]:
toronto_grouped.shape

(95, 269)

Print each neighborhood along with the top 5 most common venues.

In [36]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    

----Agincourt----
                       venue  freq
0             Clothing Store   0.2
1                     Lounge   0.2
2             Breakfast Spot   0.2
3               Skating Rink   0.2
4  Latin American Restaurant   0.2


----Alderwood / Long Branch----
          venue  freq
0   Pizza Place   0.2
1           Pub   0.1
2      Pharmacy   0.1
3           Gym   0.1
4  Skating Rink   0.1


----Bathurst Manor / Wilson Heights / Downsview North----
           venue  freq
0           Bank  0.10
1    Coffee Shop  0.10
2    Pizza Place  0.05
3          Diner  0.05
4  Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1   Chinese Restaurant  0.25
2                 Bank  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park / Lawrence Manor East----
                venue  freq
0          Restaurant  0.08
1  Italian Restaurant  0.08
2         Coffee Shop  0.08
3     Thai Restaurant  0.08
4      Sandwich Place  

4               Spa  0.02


----Regent Park / Harbourfront----
            venue  freq
0     Coffee Shop  0.15
1             Pub  0.06
2            Park  0.06
3          Bakery  0.06
4  Breakfast Spot  0.04


----Richmond / Adelaide / King----
           venue  freq
0    Coffee Shop  0.09
1           Café  0.05
2     Restaurant  0.04
3          Hotel  0.03
4  Deli / Bodega  0.03


----Rosedale----
               venue  freq
0               Park  0.50
1         Playground  0.25
2              Trail  0.25
3  Accessories Store  0.00
4      Metro Station  0.00


----Roselawn----
                             venue  freq
0                           Garden   1.0
1                Accessories Store   0.0
2               Mexican Restaurant   0.0
3  Molecular Gastronomy Restaurant   0.0
4       Modern European Restaurant   0.0


----Rouge Hill / Port Union / Highland Creek----
                             venue  freq
0                              Bar   1.0
1                Accessories Store   0.

Write a function to sort the venues in descending order.

In [41]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

 Create the new dataframe and display the top 10 venues for each neighbourhood.

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                          Agincourt                Lounge   
1                            Alderwood / Long Branch           Pizza Place   
2  Bathurst Manor / Wilson Heights / Downsview North           Coffee Shop   
3                                    Bayview Village                  Café   
4                 Bedford Park / Lawrence Manor East       Thai Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot        Clothing Store   
1                        Gym        Sandwich Place          Skating Rink   
2                       Bank    Frozen Yogurt Shop           Bridal Shop   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                Coffee Shop            Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Skating Rink   Empanada Restaurant  Ethiopian Restaurant   
1          Dance Studio              Pharmacy           Coffee Shop   
2        Sandwich Place                 Diner            Restaurant   
3           Yoga Studio        Discount Store   Distribution Center   
4    Italian Restaurant             Juice Bar              Pharmacy   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0     Electronics Store  Eastern European Restaurant     Dim Sum Restaurant  
1    Athletics & Sports                          Pub     Dim Sum Restaurant  
2           Supermarket               Ice Cream Shop       Sushi Restaurant  
3               Dog Run             Doner Restaurant             Donut Shop  
4           Pizza Place                 Cupcake Shop                    Pub

### e). Cluster Neighborhoods.

Run k-means to cluster the neighborhood into 5 clusters.

In [43]:
# set number of clusters
kclusters = 5
toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 3, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 4, 1,
       1, 1, 1, 2, 1, 1, 0], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [49]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                  

In [54]:
toronto_merged.shape

(103, 16)

Drop the rows that have no data.

In [56]:
toronto_merged=toronto_merged.dropna()
toronto_merged.shape

(100, 16)

Visualize the resulting clusters.

In [60]:
# create map
map_clusters = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### f). Examine Clusters.
Examine each cluster and determine the discriminating venue categories that distinguish each cluster.
#### Cluster 1 

In [68]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]


Neighbourhood  Cluster_Labels  \
0                                           Parkwoods               0   
21                                Caledonia-Fairbanks               0   
35                                       East Toronto               0   
66                                    York Mills West               0   
85  Milliken / Agincourt North / Steeles East / L'...               0   
91                                           Rosedale               0   
98   The Kingsway / Montgomery Road  / Old Mill North               0   

   1st Most Common Venue       2nd Most Common Venue    3rd Most Common Venue  \
0                   Park  Construction & Landscaping        Food & Drink Shop   
21                  Park                        Pool            Women's Store   
35                  Park           Convenience Store              Yoga Studio   
66                  Park                        Bank        Convenience Store   
85                  Park                  Playground               Donut Shop   
91                  Park                  Playground                    Trail   
98                  Park                       River  Comfort Food Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Yoga Studio            Donut Shop                 Diner   
21           Golf Course  Ethiopian Restaurant     Electronics Store   
35            Donut Shop                 Diner        Discount Store   
66             Drugstore        Discount Store   Distribution Center   
85    Dim Sum Restaurant                 Diner        Discount Store   
91      Doner Restaurant    Dim Sum Restaurant                 Diner   
98            Comic Shop  Ethiopian Restaurant   Empanada Restaurant   

          7th Most Common Venue        8th Most Common Venue  \
0                Discount Store          Distribution Center   
21  Eastern European Restaurant          Dumpling Restaurant   
35          Distribution Center                      Dog Run   
66                      Dog Run             Doner Restaurant   
85          Distribution Center                      Dog Run   
91               Discount Store          Distribution Center   
98            Electronics Store  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0                Dog Run       Doner Restaurant  
21             Drugstore             Donut Shop  
35      Doner Restaurant    Dumpling Restaurant  
66            Donut Shop    Dumpling Restaurant  
85      Doner Restaurant              Drugstore  
91               Dog Run             Donut Shop  
98   Dumpling Restaurant              Drugstore

#### Cluster 2 

In [69]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]


Neighbourhood  Cluster_Labels  \
1                                     Victoria Village               1   
2                           Regent Park / Harbourfront               1   
3                    Lawrence Manor / Lawrence Heights               1   
4         Queen's Park / Ontario Provincial Government               1   
7                                            Don Mills               1   
8                     Parkview Hill / Woodbine Gardens               1   
9                                      Garden District               1   
10                                           Glencairn               1   
12            Rouge Hill / Port Union / Highland Creek               1   
13                                           Don Mills               1   
14                                    Woodbine Heights               1   
15                                      St. James Town               1   
16                                  Humewood-Cedarvale               1   
17   Eringate / Bloordale Gardens / Old Burnhamthor...               1   
18                 Guildwood / Morningside / West Hill               1   
19                                         The Beaches               1   
20                                         Berczy Park               1   
22                                              Woburn               1   
23                                             Leaside               1   
24                                  Central Bay Street               1   
25                                            Christie               1   
26                                           Cedarbrae               1   
27                                   Hillcrest Village               1   
28   Bathurst Manor / Wilson Heights / Downsview North               1   
29                                    Thorncliffe Park               1   
30                          Richmond / Adelaide / King               1   
31                       Dufferin / Dovercourt Village               1   
32                                 Scarborough Village               1   
33                      Fairview / Henry Farm / Oriole               1   
34                    Northwood Park / York University               1   
36   Harbourfront East / Union Station / Toronto Is...               1   
37                           Little Portugal / Trinity               1   
38       Kennedy Park / Ionview / East Birchmount Park               1   
39                                     Bayview Village               1   
40                                           Downsview               1   
41                       The Danforth West / Riverdale               1   
42           Toronto Dominion Centre / Design Exchange               1   
43      Brockton / Parkdale Village / Exhibition Place               1   
44                   Golden Mile / Clairlea / Oakridge               1   
46                                           Downsview               1   
47                     India Bazaar / The Beaches West               1   
48                     Commerce Court / Victoria Hotel               1   
49          North Park / Maple Leaf Park / Upwood Park               1   
50                                       Humber Summit               1   
51   Cliffside / Cliffcrest / Scarborough Village West               1   
53                                           Downsview               1   
54                                     Studio District               1   
55                  Bedford Park / Lawrence Manor East               1   
56   Del Ray / Mount Dennis / Keelsdale and Silvert...               1   
57                                   Humberlea / Emery               1   
58                        Birch Cliff / Cliffside West               1   
59                                          Willowdale               1   
60                                           Downsview               1   
61                                       Lawr

#### Cluster 3 

In [70]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]


Neighbourhood  Cluster_Labels 1st Most Common Venue  \
52  Willowdale / Newtonbrook               2                   Gym   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
52          Dessert Shop           Event Space  Ethiopian Restaurant   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
52   Empanada Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
52   Dumpling Restaurant             Drugstore             Donut Shop

#### Cluster 4 

In [71]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]


Neighbourhood  Cluster_Labels 1st Most Common Venue  \
6  Malvern / Rouge               3  Fast Food Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6          Dessert Shop           Event Space  Ethiopian Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
6   Empanada Restaurant     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6   Dumpling Restaurant             Drugstore             Donut Shop

#### Cluster 5 

In [72]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]


Neighbourhood  Cluster_Labels  \
11  West Deane Park / Princess Gardens / Martin Gr...               4   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
11          Home Service           Yoga Studio            Donut Shop   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
11                 Diner        Discount Store   Distribution Center   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
11               Dog Run      Doner Restaurant             Drugstore   

   10th Most Common Venue  
11           Dessert Shop